### ISSUE TITLE
Add Bounded typeclass

### ISSUE THREAD

**Issue by julien-truffaut** [typelevel/cats#1536]
```
trait Bounded[A] extends Order[A]{
  def minValue: A
  def maxValue: A
}
```
such as `min(a, minValue) == minValue` and `max(a, maxValue) == maxValue`

**Comment by kevinmeredith** [comment]
Why must `Bounded` extend `Order`? 

I don't believe that Haskell's `Bounded` imposes such a constraint:

```haskell
iPrelude> :i Bounded
class Bounded a where
  minBound :: a
  maxBound :: a
```

**Comment by julien-truffaut** [comment]
It is correct but in this case `Bounded` will be lawless isn't it?

**Comment by kevinmeredith** [comment]
What laws must a `Bounded` instance obey?

reference - http://stackoverflow.com/questions/43029076/type-class-w-min/43029124?noredirect=1#comment73326210_43029124

**Comment by julien-truffaut** [comment]
@kevinmeredith see my comment (such as ...)

**Comment by edmundnoble** [comment]
This seems to not be `Bounded` any more when you add those laws, but `BoundedLattice` (https://github.com/typelevel/algebra/blob/master/core/src/main/scala/algebra/lattice/BoundedLattice.scala) seeing as `<` and `>` can be implemented using `meet`/`join` and `Eq`.

**Comment by johnynek** [comment]
I'm not crazy about Bounded. Would prefer LowerBounded and UpperBounded. String, for instance is LowerBounded. NonNegative numbers are LowerBounded. There are some examples of only UpperBounded as well.

**Comment by julien-truffaut** [comment]
I agree with @johnynek it would be more useful to one typeclass for upper and another for lower bound. Here is an example https://github.com/julien-truffaut/newts/blob/master/core/shared/src/main/scala/newts/Min.scala#L35

**Comment by kailuowang** [comment]
For lawless typeclasses, we might also want to consider adding them to alleycats instead.

**Comment by johnynek** [comment]
I don't see how his is lawless when it has an associated Ordering. So, they could extend Order and everything is fine, no?

**Comment by kailuowang** [comment]
@johnynek I didn't make it clear. I was referring to LowerBounded and UpperBounded.

**Comment by johnynek** [comment]
```scala 
trait LowerBounded[A] extends Order[A] {
  // law: compare(minValue, a) <= 0 for all a
  def minValue: A
}
```
I'm probably missing it, but how is that lawless? Maybe I'm missing context?

**Comment by kailuowang** [comment]
Ah for some reason I thought your upper and lower bound isn't order associated. My bad.

**Comment by johnynek** [comment]
No problem. In thinking about this more, we probably want this to extend PartialOrder so it can represent the top and bottom of any Lattice, but then maybe indeed these are equivalent to bounded meet and bounded join semi lattices as @edmundnobel suggests.

**Comment by izeigerman** [comment]
Hi everyone. I've got the following implementation of Enum and Bounded type classes here: https://github.com/izeigerman/scalanum . Is there any chance for it to be included into Cats? Is it missing anything else beside the integration with `PartialOrder`?

**Comment by Michaelt293** [comment]
I like the idea of adding these type classes to Cats. Just a couple of ideas, I think the laws could be expressed as `minValue <= a` and `maxValue >= a`. The implementation provided by @izeigerman looks good. I just wonder whether using `IndexededSeq` in `EnumIndex` is the way to go since it would then be possible to provide an empty sequence to an instance. An exception would then be thrown when using `head` and `tail` to the get the min and max values.

**Comment by LukaJCB** [comment]
Hey @izeigerman are you still interested in getting included into Cats? I'm sorry no one ever got back to you, but I think this is definitely still on the table :)

**Comment by izeigerman** [comment]
@LukaJCB, yes, I still am! Will create a PR soon.

**Comment by LukaJCB** [comment]
Awesome, thank you!

**Comment by izeigerman** [comment]
@LukaJCB Quick question though: I'm having a hard time coming up with laws for `Enum` typeclass. It doesn't appear that it has any. What makes the matter even worse is that similarly to its Haskell's counterpart, `Enum` is expected to return a runtime error under certain condition. Can you suggest the best way to handle this in `cats`? 
I was going to place `*Bounded` typeclasses into the `kernel` module, but I'm not that sure about `Enum`.

**Comment by LukaJCB** [comment]
I think we should start with adding `*Bounded` first and then discussing the specifics of `Enum` in the other ticket, #2342. :) If you need any help with setting up the laws and everything, please feel free to reach out in the `cats-dev` gitter channel!

**Comment by LukaJCB** [comment]
Closed by #2913
